sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestBindMethods.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    public interface PairRowDAO {
+
+        @SqlUpdate("INSERT INTO bind_methods (id, value) VALUES(:row.getKey, :row.getValue.getColumnValue)")
+        @RegisterArgumentFactory(BigIntNumberArgumentFactory.class)
+        int insert(@BindMethods("row") PairRow pairRow);
+
+        @SqlQuery("SELECT value FROM bind_methods WHERE id = :id")
+        Long getById(@Bind("id") int dbid);
+    }
+
+    public interface DatabaseValue<T> {
+
+        T getColumnValue();
+    }
+
+    public static final class PairRow {
+
+        private final int id;
+        private final DatabaseValue<?> value;
+
+        PairRow(final int id, final DatabaseValue<?> value) {
+            super();
+            this.id = id;
+            this.value = value;
+        }
+
+        public int getKey() {
+            return this.id;
+        }
+
+        public DatabaseValue<?> getValue() {
+            return this.value;
+        }
+    }
+
+    public static class DatabaseNumberValue<T extends Number> implements DatabaseValue<T> {
+
+        private final T value;
+
+        DatabaseNumberValue(final T value) {
+            super();
+            this.value = value;
+        }
+
+        @Override
+        public T getColumnValue() {
+            return this.value;
+        }
+    }
+
+    public static final class DatabaseLongValue extends DatabaseNumberValue<Long> {
+
+        DatabaseLongValue(final long value) {
+            super(Long.valueOf(value));
+        }
+    }
+
+    public static final class DatabaseLongValueWithOverride extends DatabaseNumberValue<Long> {
+
+        DatabaseLongValueWithOverride(final long value) {
+            super(Long.valueOf(value));
+        }
+
+        @Override
+        public Long getColumnValue() {
+            return super.getColumnValue();
+        }
+    }
+
+    public static final class DirectDatabaseLongValue implements DatabaseValue<Long> {
+
+        private final long value;
+
+        DirectDatabaseLongValue(final long value) {
+            super();
+            this.value = value;
+        }
+
+        @Override
+        public Long getColumnValue() {
+            return Long.valueOf(this.value);
+        }
+    }
+
+    public static final class BigIntNumberArgument implements Argument {
+
+        private final Number value;
+
+        public BigIntNumberArgument(final Number value) {
+            this.value = value;
+        }
+
+        @Override
+        public void apply(final int position, final PreparedStatement statement, final StatementContext ctx) throws SQLException {
+            if (this.value != null) {
+                statement.setLong(position, this.value.longValue());
+            } else {
+                statement.setNull(position, Types.BIGINT);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return Objects.toString(this.value);
+        }
+
+    }
+
+    public static final class BigIntNumberArgumentFactory implements ArgumentFactory {
```

Comment:
AbstractArgumentFactory simplifies this process by handling nulls for you--just pass the SQL type integer to the superconstructor and implement the method, which is only invoked when the value is non-null.

Reply:
After thinking about it, there is no harm in letting the argument factory handle all the cases. I've made your suggested changes and pushed an update. Thanks!